### DISCUSSION TITLE
malloc branch and arbitrarily weighted stacks generally.

### DISCUSSION THREAD

**Discussion by pnf** [async-profiler/async-profiler#642] [Ideas]
Since jmc 8.3.0 is apparently going to support flame graphs with arbitrary weight attributes, I've been playing with native memory profiling that works by emitting equal and opposite weighted stackful events on malloc/free.  So like what happens in the malloc branch, except (1) we'd do the pointer match-up before writing to jfr and (2) we could have simultaneous native allocation and cpu profiling.
In general, one might want to have other sorts of aggregable events too, and I don't want to fiddle with repeatedly inserting new BCI types in the low teens of ASGCT_CallFrameType, so it seems like it would be nice to give event type its own enum. 

I currently use jemalloc, but having the java frames (as well as the await chain per #606) would be immensely helpful.

**Comment by apangin** [comment]
I already implemented a different approach to native memory leak profiling, which obsoletes the `malloc` branch. It will get into async-profiler 3.0 once I solve the organizational issues. The main differences of the newer version are:
 - It does not require `LD_PRELOAD` and can be enabled in runtime.
 - It supports custom allocators, including jemalloc, tcmalloc, etc.
 - Works both on Linux and macOS.

It's also easy to make this work simultaneously with `cpu` profiling.

As in previous version, it records malloc/free calls as JFR events, that can be later post-processed to get the leak report. I don't think matching malloc/free addresses right inside the profiler is a good idea. First, this may require too much RAM (accounted to the JVM process), so it's better to offload the expensive job to a separate offline process. Second, having individual malloc/free events opens a way to other useful features, e.g., allocation life time analysis, heatmaps, and so on.

**Comment by pnf** [comment]
That sounds promising.  To your second point, I am also suggesting individual malloc/free events.  To the first, yes, you have to keep the malloc stack around so you can emit it with the free event, but that's at worst a (size of stack) / (byte period of sampling) effect, and generally much less, because many events share the same stack.
From my point of view, the more JMC can be "one stop shopping" for analysis, the more likely it is that my users will do any analysis.
